planner/core/logical_plan_builder.go

```diff
@@ -3559,13 +3559,52 @@ func checkUpdateList(ctx sessionctx.Context, tblID2table map[int64]table.Table,
 	if err != nil {
 		return err
 	}
+	// Collect primary keys
+	pkColumnsMap := make(map[int64]map[int64]bool)
+	for _, content := range updt.TblColPosInfos {
+		if _, ok := pkColumnsMap[content.TblID]; ok {
+			continue
+		}
+		tbl := tblID2table[content.TblID]
+		pkColumns := make(map[int64]bool)
+		var pkIdx *model.IndexInfo
+		for _, idx := range tbl.Meta().Indices {
+			if idx.Primary {
+				pkIdx = idx
+				break
+			}
+		}
+		if pkIdx == nil {
+			continue
+		}
+		for _, pkCol := range pkIdx.Columns {
+			pkColumns[tbl.Cols()[pkCol.Offset].ID] = true
+		}
+		pkColumnsMap[content.TblID] = pkColumns
+	}
+	isPKUpdated := make(map[int64]model.CIStr)
 	for _, content := range updt.TblColPosInfos {
 		tbl := tblID2table[content.TblID]
 		flags := assignFlags[content.Start:content.End]
+		var updatePK bool
 		for i, col := range tbl.WritableCols() {
 			if flags[i] && col.State != model.StatePublic {
 				return ErrUnknownColumn.GenWithStackByArgs(col.Name, clauseMsg[fieldList])
 			}
+			// Check for multi-updates on primary key,
+			// see https://dev.mysql.com/doc/mysql-errors/5.7/en/server-error-reference.html#error_er_multi_update_key_conflict
+			if !flags[i] {
+				continue
+			}
+			if col.IsPKHandleColumn(tbl.Meta()) || pkColumnsMap[content.TblID][col.ID] {
```

Comment:
Why constructing a temporary map `pkColumnsMap` to check whether `col` is primary key?
Could `col.Primary` work the same way as `pkColumnsMap[content.TblID][col.ID]` here?

Reply:
My mistake. Using `mysql.HasPriKeyFlag` is enough, and I think #20603 should also change to this. Do I have to create another PR for it?

Reply:
Yes, of course.
Don't hesitate to improve the code quality, even tiny changes.

Reply:
Sure, I've put this change into https://github.com/pingcap/tidb/pull/20493. It may take longer time to review.